### PR TITLE
Add `requestInfo` to options, to optimize database behaviour

### PR DESCRIPTION
**Pull request by TheAschr**
If you run a query like 

```
{
  users {
     totalCount
  }
}
```

the current code will fetch all users from the database for edges even though they aren't being returned to client (this is an issue for large databases).

I added an optional ```info: GraphQLResolveInfo``` argument which will not fetch edges from db if the user doesn't request them and the user is not paginating.

Also the totalCount is not queried from db if the user does not request it.

```
const resolvers = {
  Query: {
    async users(parent, args, context, info) {
      return await findManyCursorConnection(
        (cursorArgs) =>
          this.context.prisma.user.findMany({
            ...cursorArgs,
          }),
        () => this.context.prisma.user.count({}),
        {
          first: args.pagination ? args.pagination.first : undefined,
          after: args.pagination ? args.pagination.after : undefined,
          before: args.pagination ? args.pagination.before : undefined,
          last: args.pagination ? args.pagination.last : undefined,
        },
        {
          getCursor: (node) => ({ id: node.id.toString() }),
          encodeCursor: (cursor) =>
            Buffer.from(JSON.stringify({ id: cursor.id })).toString("base64"),
          decodeCursor: (cursor) =>
            JSON.parse(Buffer.from(cursor, "base64").toString("ascii")),
          info,
        }
      );
    },
  },
};
``

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -6,15 +7,17 @@ export async function findManyCursorConnection<Model = { id: string }, Cursor =
...
-  pOptions?: Options<Model, Cursor>
+  pOptions?: Options<Model, Cursor>,
 ): Promise<Connection<Model>> {
   // Make sure the connection arguments are valid and throw an error otherwise
   // istanbul ignore next
   if (!validateArgs(args)) {
     throw new Error('This code path can never happen, only here for type safety')
   }
 
+
```
Comment (queicherius) [inline src/index.ts:18]: Nit: Remove this line

File: src/index.ts [hunk 2]
```diff
@@ -6,15 +7,17 @@ export async function findManyCursorConnection<Model = { id: string }, Cursor =
...
   const options = pOptions || getDefaultOptions()
+  const topLevelFields = options.info && Object.keys(graphqlFields(options.info))
```
Comment (queicherius) [inline src/index.ts:20]: Nit: `topLevelFields` -> `requestedFields`

File: src/interfaces.ts [hunk 3]
```diff
@@ -1,8 +1,11 @@
+import { GraphQLResolveInfo } from 'graphql';
+
 // Prisma Relay Cursor Connection Arguments
 export interface Options<Model, Cursor> {
   getCursor: (node: Model) => Cursor
   encodeCursor: (cursor: Cursor) => string
-  decodeCursor: (cursorString: string) => Cursor
+  decodeCursor: (cursorString: string) => Cursor,
+  info?: GraphQLResolveInfo;
```
Comment (queicherius) [inline src/interfaces.ts:8]: Nit: `info` -> `resolveInfo`

File: src/index.ts [hunk 4]
```diff
@@ -63,8 +68,11 @@ export async function findManyCursorConnection<Model = { id: string }, Cursor =
     if (hasPreviousPage) nodes.shift()
   } else {
     // Execute the underlying query operations
-    nodes = await findMany({})
-    totalCount = await aggregate()
+    nodes = await findMany({
+      take: topLevelFields === undefined || topLevelFields.includes('edges') ? undefined : 0,
```
Comment (queicherius) [inline src/index.ts:72]: Add a small explanatory comment here:

`// If the edges were not in the request fields, do not load any nodes`

**Review by queicherius — CHANGES_REQUESTED** [review]

**Comment by queicherius** [comment]
Thank you for the PR!

> The current code will fetch all users from the database for edges
> even though they aren't being returned to client (this is an issue for large databases).

Just a small headsup: You should likely never allow client to read out the entire database (even if they define "edges"). (I personally added this case without pagination so I can use the package internally as well.) See this hint in the readme:

> Two cases need to be checked in your code if you are passing in user-provided data to prevent the user from reading out too many resources at once:
> One of first | last has to be defined

That said, I think this is still a worthwhile optimization / use-case.

I left you some review comments, in addition could you please:

- Add tests, to get back to 100% coverage
- Add a small documentation block, below the [Custom Cursors](https://github.com/devoxa/prisma-relay-cursor-connection#custom-cursors) section (as it's own new section)

**Comment by queicherius** [comment]
Hey! Please let me know if you are still interested in getting this in.

**Comment by Nedomas** [comment]
Just wanted to followup on this regarding the status and if there's any help needed here. I have a case where I see `count` queries are being executed when `totalCount` is not present in the query. Would also like for it to only query `count` if its included in the query. 

I understand this PR could fix this? What would be the next steps for getting this released then?

**Comment by queicherius** [comment]
@Nedomas 

> I have a case where I see count queries are being executed when totalCount is not present in the query. Would also like for it to only query count if its included in the query. I understand this PR could fix this?

Yes, that is correct.

> Just wanted to followup on this regarding the status and if there's any help needed here. What would be the next steps for getting this released then?

This PR is stale right now, you can see my review above (a few small nits, and missing tests & documentation). If you would like to get this in feel free to use this code as the basis and send in another PR with those points resolved, and I will happily review it.

File: src/interfaces.ts (hunk 3 above)
Comment (Nedomas) [reply src/interfaces.ts:8]: This param is usually called `info` (see what prisma v1 used to have). Not sure if anything changed, but I'd rather not have to do `resolveInfo: info` and instead be able to do `{ info }` here. Let me know what you think

**Comment by queicherius** [comment]
I am going to close this PR since it's been open for ~9 months without any change now and it has drifted from `master`. If someone wants to pick it back up feel free to open a fresh PR.

**Pull request closed without merging**